Forge Networking Remastered Unity/Assets/Bearded Man Studios Inc/Scripts/Networking/Forge/Networking/FacepunchP2PClient.cs

```diff
@@ -0,0 +1,430 @@
+		{
+			RoomEnter roomEnter = await lobbyToJoin.Join();
+			if (roomEnter != RoomEnter.Success)
+				return;
+			Lobby = lobbyToJoin;
+			Connect(Lobby.Owner.Id, pendCreates);
+		}
+
+		/// <summary>
+		/// Disconnect this client from the server
+		/// </summary>
+		/// <param name="forced">Used to tell if this disconnect was intentional <c>false</c> or caused by an exception <c>true</c></param>
+		public override void Disconnect(bool forced)
+		{
+			Logging.BMSLog.Log("<color=cyan>FacepunchP2P client disconnecting...</color>");
+
+			if (Lobby.Id.Value > 0)
+				Lobby.Leave();
+
+			if (Client == null)
+				return;
+
+			lock (Client)
+			{
+				if (forced)
+					CloseConnection();
+				else
+				{
+					var frame = new ConnectionClose(Time.Timestep, false, Receivers.Server, MessageGroupIds.DISCONNECT, false);
+					Send(frame, true);
+					Task.Queue(CloseConnection, 1000);
+				}
+
+				// Send signals to the methods registered to the disconnect events
+				if (forced)
+					OnForcedDisconnect();
+			}
+		}
+
+		/// <summary>
+		/// Infinite loop listening for new data from all connected clients on a separate thread.
+		/// This loop breaks when readThreadCancel is set to true
+		/// </summary>
+		private void ReadNetwork()
+		{
+			var messageFrom = default(SteamId);
+
+			try
+			{
+				BMSByte packet = null;
+				// Intentional infinite loop
+				while (IsBound)
+				{
+					// If the read has been flagged to be canceled then break from this loop
+					if (readThreadCancel)
+						return;
+
+					try
+					{
+						if (SteamNetworking.IsP2PPacketAvailable())
+						{
+							packet = Client.Receive(out messageFrom);
+						}
+						else
+						{
+							Thread.Sleep(1);
+							continue;
+						}
+
+						if (messageFrom == default(SteamId))
+						{
+							Thread.Sleep(1);
+							continue;
+						}
+
+						if (packet == null)
+						{
+							Logging.BMSLog.LogWarning("Null packet received from player: " + messageFrom.Value);
+							Thread.Sleep(1);
+							continue;
+						}
+
+						if (PacketLossSimulation > 0.0f && new Random().NextDouble() <= PacketLossSimulation)
+						{
+							// Skip this message
+							continue;
+						}
+
+						BandwidthIn += (ulong)packet.Size;
+					}
+					catch (SocketException ex)
+					{
+						// This is a common exception when we exit the blocking call
+						Logging.BMSLog.LogException(ex);
+						Disconnect(true);
+					}
+
+					// Check to make sure a message was received
+					if (packet == null || packet.Size <= 0)
+						continue;
+
+					// Check to see if the headers have been exchanged
+					if (!headerExchanged)
+					{
+						if (Websockets.ValidateResponseHeader(headerHash, packet.CompressBytes()))
+						{
+							headerExchanged = true;
+
+							// TODO:  When getting the user id, it should also get the server time
+							// by using the current time in the payload and getting it back along with server time
```

Comment:
Was this TODO copied over? Might be worth removing if yes.

Reply:
Yes, and yes.